File: 141.环形链表.js
/*
 * @lc app=leetcode.cn id=141 lang=javascript
 *
 * [141] 环形链表
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */

/**
 * @param {ListNode} head
 * @return {boolean}
 */
// 使用快慢指针判断环的存在
var hasCycle = function(head) {
    if (head == null || head.next == null) return false
    
    let slow = head, fast = head.next
    while (slow != fast) {
        if (fast == null || fast.next == null) return false
        slow = slow.next
        fast = fast.next.next
    }
    return true
}

// @lc code=end
// 使用哈希表存储
var hasCycle_1 = function(head) {
    let map = new Map()
    let cur = head
    while(cur && !map.has(cur)) {
        map.set(cur, 0)
        cur = cur.next
    }
    return cur != null
};

File: 141.环形链表.java
import java.util.HashSet;
import java.util.Set;

/*
 * @lc app=leetcode.cn id=141 lang=java
 *
 * [141] 环形链表
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
// public class ListNode {
//     int val;
//     ListNode next;

//     ListNode(int x) {
//         val = x;
//     }
// }
public class Solution {
    public boolean hasCycle1(ListNode head) {
        if (head == null) return false;
        ListNode slow = head, fast = head;
        while(fast.next != null && fast.next.next != null) {
            fast = fast.next.next;
            slow = slow.next;
            if (fast == slow) {
                return true;
            }
        }
        return false;
    }
    public boolean hasCycle(ListNode head) {
        if (head == null) return false;
        ListNode cur = head;
        HashSet<ListNode> set = new HashSet<>();
        while(cur != null) {
            if (set.contains(cur)) return true;
            set.add(cur);
        }
        return false;
    }
    // public static void main(String[] args) {
    //     Solution s = new Solution();
    //     ListNode n1 = new ListNode(3);
    //     boolean a = s.hasCycle(n1);
    //     System.out.println(a);
    // }
}
// @lc code=end

